### ISSUE TITLE
redundant_optional_initialization does not find properties with observers (e.g. didSet) 

### ISSUE THREAD

**Issue by iressler** [realm/SwiftLint#3621]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

The redundant_optional_initialization rule does not trigger when a property has observers (e.g. didSet, willSet), but will without them.

For example this property will trigger redundant_optional_initialization:
```swift
var varName: String? = nil
```

but this property will not:
```swift
var varName: String? = nil {
  didSet {
    print("didSet")
  }
}
```

##### Complete output when running SwiftLint, including the stack trace and command used

```zsh
$ swiftlint lint
Loading configuration from '.swiftlint.yml'
Linting Swift files at paths
Linting 'SwiftLintTestApp.swift' (1/2)
Linting 'ContentView.swift' (2/2)
/Users/iressler/repos/SwiftLintTest/SwiftLintTest/ContentView.swift:11:23: warning: Redundant Optional Initialization Violation: Initializing an optional variable with nil is redundant. (redundant_optional_initialization)
Done linting! Found 1 violation, 0 serious in 2 files.
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)?
```zsh
swiftlint version
0.43.1
```
* Installation method used (Homebrew, CocoaPods, building from source, etc)?
CocoaPods, Homebrew, and install from source all have the same result.
* Paste your configuration file:
This occurs with an empty configuration file.

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)?
No.
* Which Xcode version are you using (check `xcodebuild -version`)?
```zsh
xcodebuild -version
Xcode 12.4
Build version 12D4e
```

* Do you have a sample that shows the issue?

This example properly triggers redundant_optional_initialization:
```zsh
echo "var varName: String? = nil" | swiftlint lint --no-cache --use-stdin
<nopath>:1:21: warning: Redundant Optional Initialization Violation: Initializing an optional variable with nil is redundant. (redundant_optional_initialization)
Done linting! Found 1 violation, 0 serious in 1 file.
```

This example does not trigger it:
```zsh
echo "var varName: String? = nil { didSet { print("something") } }" | swiftlint lint --no-cache --use-stdin
Done linting! Found 0 violations, 0 serious in 1 file.
```

**Comment by iressler** [comment]
If this is a real bug and not intentional I'd be willing to take a stab at figuring it out/fixing it. I've poked at the project a couple times, but not enough to say I understand much.

**Comment by marcelofabri** [comment]
Yeah, I consider it a real bug. Thanks for volunteering to take a look at this, @iressler! Let me know if you need any help or guidance.

**Comment by iressler** [comment]
I finally got a chance to dig into this, and here is what I found.

The false negative occurs on line 126 of `RedundantOptionalInitializationRule` where it checks the range of a regex match:
```Swift
match.location == range.location + range.length - match.length
```

I assume the comparison is trying to validate that the match is at the end of the range, which fails in this case because the range includes the ` {` for the observer(s).

I created a PR (#3671) that has a proposed change to the regex. I'm not sure if it's the best change, but I'm hoping it will at least keep the conversation going.

**Comment by iressler** [comment]
@marcelofabri I'd like to get your thoughts on what I found/my suggested change. If you think there's a different approach I'd be happy to give it a shot.

**Comment by marcelofabri** [comment]
Implemented in #3671

**Comment by jonas-simkus-mw** [comment]
This is not fixed, still an issue so many years later.
